### PR TITLE
[FLINK-6578] Fix self-loop handling in SharedBuffer.

### PR DESCRIPTION
**Pull request by kl0u**
This RP addresses: FLINK-6604, FLINK-6371, FLINK-6536, FLINK-6255, and FLINK-6578.

R @dawidwys

### CODE REVIEW

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 1]
```diff
@@ -593,15 +596,24 @@ private boolean checkFilterCondition(ComputationState<T> computationState, Itera
 		return condition == null || condition.filter(event, computationState.getConditionContext());
 	}
 
+	/**
+	 * Extracts all the sequences of events from the start to the given computation state. An event
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: With the `Preconditions.checkState(paths.size() <= 1);` all = 1

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 2]
```diff
@@ -83,26 +84,29 @@ public SharedBuffer(final TypeSerializer<V> valueSerializer) {
 	 * @param previousTimestamp Timestamp of the value for the previous relation
 	 * @param version           Version of the previous relation
 	 */
-	public void put(
+	public int put(
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: missing javadoc for return

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 3]
```diff
@@ -114,16 +118,16 @@ public void put(
 	 * @param timestamp Timestamp of the current value (a value requires always a timestamp to make it uniquely referable))
 	 * @param version   Version of the previous relation
 	 */
-	public void put(
+	public int put(
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: missing javadoc for return

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 4]
```diff
@@ -256,8 +271,8 @@ public void prune(long pruningTimestamp) {
 	 * @param value     Value to lock
 	 * @param timestamp Timestamp of the value to lock
 	 */
-	public void lock(final K key, final V value, final long timestamp) {
-		SharedBufferEntry<K, V> entry = get(key, value, timestamp);
+	public void lock(final K key, final V value, final long timestamp, int counter) {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: missing javadoc for counter

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 5]
```diff
@@ -271,8 +286,8 @@ public void lock(final K key, final V value, final long timestamp) {
 	 * @param value     Value to release
 	 * @param timestamp Timestamp of the value to release
 	 */
-	public void release(final K key, final V value, final long timestamp) {
-		SharedBufferEntry<K, V> entry = get(key, value, timestamp);
+	public void release(final K key, final V value, final long timestamp, int counter) {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: missing javadoc for counter

**Comment by kl0u** [comment]
Merged.

**Pull request closed without merging**